### PR TITLE
[2D parallel] workaround for FSDP init issue

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103866

Closes https://github.com/pytorch/pytorch/issues/96491 and does so by relaxing FSDP's assumption that the entire input module must be on the same device. Now, FSDP can accept a module partially on CPU and GPU and just emits a warning.
Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

### CODE REVIEW

File: torch/distributed/fsdp/_init_utils.py [hunk 1]
```diff
@@ -753,6 +753,12 @@ def _check_single_device_module(
     module must be either fully on the CPU or fully on a non-CPU device.
     """
     devices = {param.device for param in _get_orig_params(module, ignored_params)}
+    if len(devices) == 2 and torch.device("cpu") in devices:
+        warnings.warn(
+            f"FSDP wrapping module with params on multiple devices: {devices}"
+            "This could occur in cases where another parallelism algorithm has already"
+            "moved some parameters to a non-CPU device."
+        )
     if len(devices) > 1:
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:762]: If `if len(devices) == 2 and torch.device("cpu") in devices == True`, I am not seeing why this `if len(devices) > 1:` branch would not also get taken, still leading to the `RuntimeError` getting raised.

Comment (rohan-varma) [reply torch/distributed/fsdp/_init_utils.py:762]: yeah, we need to change this to an `else`

File: torch/distributed/fsdp/_init_utils.py [hunk 2]
```diff
@@ -746,14 +746,34 @@ def _get_buffer_names(root_module: nn.Module) -> Set[str]:
...
+    # to not be None, otherwise we'd flatten parameters in a mixed module which is
+    # not supported.
+    if (
+        len(devices) == 2 and torch.device("cpu") in devices
+    ):
+        if device_id is not None:
+            warnings.warn(
+                f"FSDP wrapping module with params on multiple devices: {devices}"
+                "This could occur in cases where another parallelism algorithm has already"
+                "moved some parameters to a non-CPU device."
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:770]: I am not sure if we should give this warning.

What I am thinking is that one of the following cases holds:
- The user knows that he/she is using mixture of parallelisms, and then this warning is not too informative.
- The user does not know that he/she is using a mixture of parallelisms, and then this warning may be jarring.
- the user is not using a mixture of parallelisms but hits this branch somehow, and then this warning may be confusing.

File: torch/distributed/fsdp/_init_utils.py [hunk 3]
```diff
@@ -746,14 +746,34 @@ def _get_buffer_names(root_module: nn.Module) -> Set[str]:
...
+            )
+        else:
+            raise RuntimeError(
+                f"To support partial CPU / GPU module, please pass in device_id argument."
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:774]: nit: Maybe "To support a module with both CPU and GPU parameters"?

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 4]
```diff
@@ -482,20 +482,53 @@ def test_module_device_mismatches_device_id(self):
...
+            RuntimeError, "please pass in device_id"
+        ):
+            FSDP(CPUGPUModule())
+
+    @skip_if_lt_x_gpu(2)
+    def test_multigpu_module(self):
+        """
+        Module on multiple GPUs wrapped in FSDP should raise an error.
+        """
+        torch.cuda.set_device(self.rank)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:515]: curious: Do we not already set this in the test setup? (I feel like we rely on it being set.)

**Review by awgu — APPROVED** [review]

File: test/distributed/tensor/parallel/test_2d_parallel.py [hunk 5]
```diff
@@ -50,22 +51,20 @@ def _distribute_and_fsdp_wrap_module(
...
-            module.net1, process_group=pg, use_orig_params=use_orig_params
-        )
-        module.net2 = FSDP(
-            module.net2, process_group=pg, use_orig_params=use_orig_params
-        )
-    return FSDP(module, process_group=pg, use_orig_params=use_orig_params)
+        module.net1 = fsdp_ctor(module.net1)
+        module.net2 = fsdp_ctor(module.net2)
+
+    return fsdp_ctor(module)
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_2d_parallel.py:59]: Thanks for cleaning this up.

File: test/distributed/tensor/parallel/test_2d_parallel.py [hunk 6]
```diff
@@ -50,22 +51,20 @@ def _distribute_and_fsdp_wrap_module(
...
 
 
 def init_model(model_parallel_size=TP_DEGREE, use_orig_params=False, fsdp_nested=False):
     rank = dist.get_rank()
     torch.cuda.set_device(rank)
     world_size = dist.get_world_size()
 
-    model = SimpleModel().cuda(rank)
+    model = SimpleModel()
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_2d_parallel.py:67]: Nice!

**Review by fduwjj — APPROVED** [review]

**Pull request closed without merging**